cmd-ref: document new gc behavior

skshetry opened the pull request:
> You may disregard these recommendations if you used the **Edit on GitHub** button from dvc.org to improve a doc in place.

❗ Please read the guidelines in the [Contributing to the Documentation](https://dvc.org/doc/user-guide/contributing/docs) list if you make any substantial changes to the documentation or JS engine.

🐛 Please make sure to mention `Fix #issue` (if applicable) in the description of the PR. This causes GitHub to close it automatically when the PR is merged.

Please chose to [allow us](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to edit your branch when creating the PR.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Fixes #1013

shcheklein commented on public/static/docs/command-reference/gc.md:
```diff
@@ -42,16 +46,20 @@ restored using `dvc fetch`, as long as they have previously been uploaded with
   useful if tags are used to track "checkpoints" of an experiment or project.
   Note that both options can be combined, for example using the `-aT` flag.
 
+- `-w`, `--workspace` - remove files in local cache. _This operation is
```
may be I'm missing something, but `workspace` is not the opposite to `-c`. It's one of the options (along with `-a` and `-T`) to provide the scope of files we would like to keep (and remove everything else). So this option description should be similar to `-a` and/or `-T`

skshetry replied:
`-w` triggers the same behavior as plain-old `gc`, in that it deletes thing (unless `-a`/`-T` is specified). It's similar to `-c` option..

shcheklein replied:
from the description it sounds like it does delete all files in cache. This is not true (I hope). We need to describe it somehow. The best to describe is to use "keep" semantics like in `-a` or `-T` (i.e. what scope of DVC-file references we keep and remove everything else in cache).

Also, I hope I can run just `-a` still, right? Or just `-T`?

skshetry replied:
Ahh, okay. I misunderstood you there.

skshetry replied:
Yes, you can just do `-a` or `-T` without mentioning `-w`.

shcheklein commented on public/static/docs/command-reference/gc.md:
````diff
@@ -82,7 +90,7 @@ When you run `dvc gc` it removes all objects from cache that are not referenced
 in the <abbr>workspace</abbr> (by collecting hash values from the DVC-files):
 
 ```dvc
-$ dvc gc
````
change the paragraph above, please

skshetry replied:
Fixed.

shcheklein commented on public/static/docs/command-reference/gc.md:
````diff
@@ -5,7 +5,7 @@ Remove unused objects from <abbr>cache</abbr> or remote storage.
 
 ```usage
-usage: dvc gc [-h] [-q | -v] [-a] [-T] [-c] [-r REMOTE] [-f] [-j JOBS]
+usage: dvc gc [-h] [-q | -v] [-w] [-a] [-T] [-c] [-r REMOTE] [-f] [-j JOBS]
````
minor: let's split into two lines this one (three lines total?), it makes sense at this point (everything after aux options h, q, v can go to the next line)

jorgeorpinel commented on public/static/docs/command-reference/gc.md:
```diff
@@ -14,9 +15,13 @@ usage: dvc gc [-h] [-q | -v] [-a] [-T] [-c] [-r REMOTE] [-f] [-j JOBS]
-the local cache, which is typically located on the same machine as the project
-in question. This usually helps to free up disk space.
+currently in the <abbr>workspace</abbr>. By default, this command won't delete
+anything at all to make it safe and explicit. However, you can use different
```
> By default, this command won't delete anything at all.

Is this true? `dvc gc` doesn't do anything? Does it output a warning about this at least?

skshetry replied:
Yes, it doesn't *do* anything. And, it does show error message if tried without any flags. Though for the docs, this is not important.

jorgeorpinel replied:
Thanks for the info. I'm not so sure this is unimportant for docs but will follow up in https://github.com/iterative/dvc.org/pull/1023#discussion_r386595996 below. Resolving 🙂

jorgeorpinel commented:
Question: do -a and -T keep also cached files referenced in the workspace regardless of whether you're currently checked out in a branch or tag? If so we should update those options to explain so, something like "in all Git branches/tags as well as the workspace".

jorgeorpinel commented on public/static/docs/command-reference/gc.md:
```diff
@@ -42,16 +47,20 @@ restored using `dvc fetch`, as long as they have previously been uploaded with
-  the current commit (unless `-a` or `-T` are also used). The default remote is
-  used unless a specific one is given with `-r`.
+- `-c`, `--cloud` - remove files in remote storage in addition to local cache.
+  _This operation is dangerous._ It removes datasets, models, other files that
```
```suggestion
  **This behavior is dangerous.** It removes datasets, models, or other files that
```

jorgeorpinel commented on public/static/docs/command-reference/gc.md:
```diff
@@ -78,11 +87,12 @@ $ du -sh .dvc/cache/
 
-When you run `dvc gc` it removes all objects from cache that are not referenced
-in the <abbr>workspace</abbr> (by collecting hash values from the DVC-files):
+When you run `dvc gc --workspace` it removes all objects from cache that are not
```
Minor: I'd say let's use the shorthand -w for the example?
```suggestion
When you run `dvc gc -w`, DVC removes all objects from cache that are not
```

skshetry replied:
I don't know what's more used in the docs, but, I'd say we should prefer long form flags (they are pretty self-explanatory) wherever possible. Or, unless we want to show combinations of flag like following:
```sh
$ dvc gc -caT  # 🐈
```

I'd leave to the user to use/find short-form flags.

jorgeorpinel replied:
No strong opinion, up to you.

jorgeorpinel replied:
But lets change "it" for "DVC" at least 🙂

jorgeorpinel commented on public/static/docs/command-reference/gc.md:
````diff
@@ -78,11 +87,12 @@ $ du -sh .dvc/cache/
 
 ```dvc
-$ dvc gc
+$ dvc gc --workspace
````
```suggestion
$ dvc gc -w
```

skshetry commented on public/static/docs/command-reference/gc.md:
```diff
@@ -14,9 +15,13 @@ usage: dvc gc [-h] [-q | -v] [-a] [-T] [-c] [-r REMOTE] [-f] [-j JOBS]
+usually helps to free up disk space.
 
 There are important things to note when using Git to version the
 <abbr>project</abbr>:
```
> completely collect the local cache

It's an implementation detail.

jorgeorpinel replied:
Not really, it explains the behavior of `dvc gc -w`.
I was just rewording "only clean up the local cache" in a clearer way.

skshetry replied:
I don't think it matters to users what we collect. "We keep everything what's currently used in workspace and remove everything else" is a better explanation imo.

jorgeorpinel replied:
The command is called "garbage collect". I think it's pretty key to explain what it collects with each option 🙂  Again, I was just rewording the previous phrase "only clean up the local cache" (it already explained what is collected). But definitely, if you have a better suggestion go for it.

> "We keep everything what's currently used in workspace and remove everything else"

This is also explaining what is (not) collected. Anyway, yeah it sounds good but probably won't make sense in the current context unless you do other changes, feel free to try using that and read the paragraphs around it to check the text still makes sense.

skshetry replied:
@jorgeorpinel, I haven't changed regarding `it will completely collect the local cache`. Rest of the suggestions were followed.

skshetry commented on public/static/docs/command-reference/gc.md:
```diff
@@ -42,16 +47,20 @@ restored using `dvc fetch`, as long as they have previously been uploaded with
 
+- `-w`, `--workspace` - remove files in local cache that are not referenced in
+  the workspace. _This operation is dangerous._ If `--all-tags` or
+  `--all-branches` are specified, this flag is not explicitly required.
```
No, they can work with `-w` or `-c`, so, it kind-of makes sense to also mention there. Also, `-a` or `-T` by default activates -`w`.

jorgeorpinel replied:
Are we sure? Can you answer https://github.com/iterative/dvc.org/pull/1023#issuecomment-593587976 please? Copied here:

> Question: do -a and -T keep also cached files referenced in the workspace regardless of whether you're currently checked out in a branch or tag?

I find this mention about -aT here pretty confusing so we need to double check that to figure out the best phrasing (or remove). At this time I think moving this option to the top is probably enough (no mention about -aT, but probably mention -w somehow in -aT.

skshetry replied:
Yes, when  `-a` and `-T` are provided, it is same as `-wa` and `-wT` respectively, so, it still keeps everything from workspace.

> Question: do -a and -T keep also cached files referenced in the workspace regardless of whether you're currently checked out in a branch or tag?

Yes, they should (but, I'd verify it tomorrow and update).

jorgeorpinel replied:
> when -a and -T are provided, it is same as -wa and -wT respectively

OK, I just double checked (DVC 0.87.0) when the worskspace is in a detached head (no branch or tag) and indeed this seems to be the case. In fact `dvc gc -a` warns: `WARNING: This will remove all cache except items used in the working tree and all git branches of the current repo.`

so to summarize, my suggestion is to move -w to the top of the options list, and something like
```suggestion
  the workspace. **This behavior is dangerous.** This option is
  enabled automatically if `--all-tags` or `--all-branches` are
  used.
```
Thanks!

> I may open a separate suggestion about how to mention -w in -aT.

jorgeorpinel commented on public/static/docs/command-reference/gc.md:
```diff
@@ -14,9 +15,13 @@ usage: dvc gc [-h] [-q | -v] [-a] [-T] [-c] [-r REMOTE] [-f] [-j JOBS]
-in question. This usually helps to free up disk space.
+currently in the <abbr>workspace</abbr>. By default, this command won't delete
+anything at all to make it safe and explicit. However, you can use different
+flags to change the behavior.
```
Maybe `anything at all (a warning is shown),` but yeah, maybe that's not relevant, up to you.

jorgeorpinel replied:
I checked on DVC 0.87.0 and actually running without flags causes an ERROR output (`ERROR: unexpected error - Invalid Arguments. Either of (all_branches, all_tags, all_commits, workspace, cloud) needs to be enabled.`). Maybe we should express this differently like:
```suggestion
currently in the <abbr>workspace</abbr>. To avoid accidentally deleting data, this command
requires the explicit use of explicit [options](#options) to determine it's behavior (what
"garbage" to collect).
```
> Prob needs wraping.

shcheklein commented:
@skshetry @jorgeorpinel what is the status of it? @skshetry is still assigned - are you going to move it forward?

skshetry commented:
@shcheklein Umm, yes. I'm trying to add `--all-commits` on this PR as well. I'll also have to address @jorgeorpinel reviews.

shcheklein commented:
@skshetry sure! np. Just wanted to be sure that I don't block it like it happened last time.

shcheklein commented:
@jorgeorpinel @skshetry ping! any progress on this?

skshetry commented:
@shcheklein, sorry, I couldn't get to the docs. I have adjusted it now and addressed most of the suggestions. :slightly_smiling_face: 

Please take a look.

shcheklein commented on public/static/docs/command-reference/gc.md:
```diff
@@ -14,17 +15,25 @@ usage: dvc gc [-h] [-q | -v] [-a] [-T] [-c] [-r <name>]
 
-There are important things to note when using Git to version the
-<abbr>project</abbr>:
+By default, this command won't delete anything at all to make it safe and
```
looks like a repetition to the previous sentence

skshetry replied:
fixed, thanks

shcheklein commented:
@skshetry @jorgeorpinel PTAL - https://dvc-landing-gc-flag-cha-stdoiq.herokuapp.com/doc/command-reference/gc - I've applied some edits to the doc

skshetry commented on content/docs/command-reference/gc.md:
```diff
@@ -1,60 +1,75 @@
-the local cache, which is typically located on the same machine as the project
-in question. This usually helps to free up disk space.
+This command deletes (garbage collects) data files or directories that exist in
+DVC cache but are no longer needed. With `--cloud` it also removes data in
```
> but are no longer needed.

It'd be better to express this in any other way. I don't have a better suggestion, though than the following:
```suggestion
DVC cache but are no longer in use. With `--cloud` it also removes data in
```

shcheklein replied:
Good question, I think needed was there before. I don't have a very strong opinion ... may be even `needed` better because it is more precise here - I just keep stuff that I need. need is more general then use. Still, need to review other place with in use/need.

shcheklein replied:
Changed to always use `need` vs `in-use`. I think `need` and `needed` are actually more correct here.

jorgeorpinel commented on content/docs/command-reference/gc.md:
```diff
@@ -1,60 +1,75 @@
-  current one will be deleted.
-- Use the `--all-branches` or `--all-tags` options to avoid collecting data
-  referenced in the tips of all branches or all tags, respectively.
+One of the scope options - `--workspace`, `--all-branches`, `--all-tags`, or
```
```suggestion
One of the scope options, `--workspace`, `--all-branches`, `--all-tags`,
```

jorgeorpinel commented on content/docs/command-reference/gc.md:
```diff
@@ -1,60 +1,75 @@
+`--all-commits`, or any combination of them must be provided. Each of them
+corresponds to the current workspace _and_ a set of commits to analyze what
+files, directories and what versions are still needed and should be kept (by
+analyzing DVC-files in those commits).
```
```suggestion
`--all-commits`, or any combination of them must be provided. Each of them
corresponds to keeping the data for the current workspace and for a
different set of commits (determined by reading the DVC-files in them).
```

jorgeorpinel replied:
Question: what's the point of this option? To me it sounds like "don't gc anything" except maybe deleted parts of the repo like removed branches, rewritten commits, etc.

shcheklein replied:
DVC cache contains data that is not referenced in any way in Git history. E.g. run do a lot of `dvc run`, or `dvc repro` after modifying certain files/settings. If `--no-commit` is not specified DVC is saving data to cache not matter you do `git commit` to save DVC-files or not.

shcheklein replied:
We probably mention this.

shcheklein replied:
mentioned a possible use case in the recent version - @jorgeorpinel pls take a look

jorgeorpinel replied:
Makes sense, thanks.

jorgeorpinel replied:
Still this suggestion wasn't applied but I'll address in regular updates.

shcheklein replied:
I put a new paragraph that explains the use case, have you seen it? Or may be I didn't get you specific suggestion ... 🤔

jorgeorpinel replied:
I see it, thanks. May reword a little in regular updates but it's great!

> One of the use cases for this option is to safely delete all temporary data DVC cached when dvc run and/or dvc repro were run without committing changes to DVC-files (thus potentially caching data that is not referenced from workspace or Git commits).

jorgeorpinel commented on content/docs/command-reference/gc.md:
```diff
@@ -1,60 +1,75 @@
+
+- `-a`, `--all-branches` - keep cached objects referenced in all Git branches as
+  well as in the workspace (implies `-w`). Useful for keeping data for all the
+  latest experiment versions if branches are used to track those.
```
```suggestion
  well as in the workspace (implies `-w`). Useful if branches are used to
  track
  different experiments.
```

jorgeorpinel commented on content/docs/command-reference/gc.md:
```diff
@@ -1,60 +1,75 @@
+  latest experiment versions if branches are used to track those.
+
+- `-T`, `--all-tags` - the same as `-a` above, but applies to Git tags as well
+  as in the workspace (implies `-w`). Useful if tags are used to track
```
```suggestion
  as the workspace (implies `-w`). Useful if tags are used to track
```

jorgeorpinel commented on content/docs/command-reference/gc.md:
```diff
@@ -1,60 +1,75 @@
-  Note that both options can be combined, for example using the `-aT` flag.
+- `--all-commits` - the same as `-a` or `-T` above, but applies to Git commits
+  as well as in the workspace (implies `-w`). Useful for keeping data for all
+  experiment versions ever used in the history of the project.
```
```suggestion
  as well as the workspace (implies `-w`). Useful for keeping all the data
  used in the entire existing commit history of the project.
```

shcheklein commented:
@jorgeorpinel @skshetry PTAL again.

shcheklein commented:
@jorgeorpinel merging this, feel free to do changes on top of it!

jorgeorpinel commented:
Great 🎉  Just added a couple more changes in 0f10925e (part of #1072)
